contracts/src/NativeTokenDestination.sol

```diff
@@ -243,62 +252,105 @@ contract NativeTokenDestination is
+                emit CollateralAdded({amount: amount, remaining: updatedReserveImbalance});
                 adjustedAmount = 0;
                 currentReserveImbalance = updatedReserveImbalance;
             }
         }
 
-        // Emit an event even if the amount is zero to improve traceability.
-        emit NativeTokensMinted(recipient, adjustedAmount);
-
-        // Only call the native minter precompile if we are minting any coins.
-        if (adjustedAmount > 0) {
-            totalMinted += adjustedAmount;
-            // Calls NativeMinter precompile through INativeMinter interface.
-            NATIVE_MINTER.mintNativeCoin(recipient, adjustedAmount);
-        }
+        // Call {_mint} even if the adjustedAmount is 0 to improve traceability.
+        _mint(recipient, adjustedAmount);
     }
 
     /**
      * @dev See {TeleporterTokenDestination-_burn}
      */
-    function _burn(uint256 amount) internal virtual override {
+    function _burn(uint256 amount) internal override {
         // Burn native token by transferring to BURN_FOR_TRANSFER_ADDRESS
-        token.transfer(BURN_FOR_TRANSFER_ADDRESS, amount);
+        token.safeTransfer(BURN_FOR_TRANSFER_ADDRESS, amount);
     }
 
     /**
-     * @dev See {TeleporterTokenDestination-_scaleTokens}
+     * @dev See {TeleporterTokenDestination-_handleSendAndCall}
+     *
+     * Mints the tokens to this contract, and send them to the recipient contract as a
+     * part of the call to {INativeSendAndCallReceiver-receiveTokens} on the recipient contract.
+     * If the call fails, the amount is sent to the fallback recipient.
+     *
+     * Note: If the recipient contract does not properly handle the full msg.value sent,
+     * the balance can be locked in the recipient contract. Receiving contracts must make
+     * sure to properly handle the balance to ensure it does not get locked improperly.
      */
-    function _scaleTokens(uint256 value, bool isReceive) internal view override returns (uint256) {
-        // Multiply when multiplyOnReceive and isReceive are both true or both false.
-        if (multiplyOnReceive == isReceive) {
-            return value * tokenMultiplier;
+    function _handleSendAndCall(
+        SingleHopCallMessage memory message,
+        uint256 amount
+    ) internal override {
+        // If the contract is not yet fully collateralized, the use of send and call is not allowed
+        // because it could result in unexpected behavior given that the amount of tokens used to make the
+        // call to "receiveTokens" is less than expected. Instead, the amount is handled as a normal bridge
+        // event to fallback recipient.
+        if (!_isCollateralized()) {
+            _withdraw(message.fallbackRecipient, amount);
+            return;
         }
-        // Otherwise divide.
-        return value / tokenMultiplier;
+
+        // Mint the tokens to this contract address.
+        NATIVE_MINTER.mintNativeCoin(address(this), amount);
```

Comment:
Should we be calling `_mint` here? Here we're missing incrementing `totalMinted`